packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts

```diff
@@ -66,60 +56,89 @@ export const useGridPagination = (apiRef: GridApiRef): void => {
 
       setGridState((state) => ({
         ...state,
-        pagination: applyConstraints(
-          {
-            ...state.pagination,
-            pageSize,
-          },
-          options.page,
-        ),
+        pagination: applyConstraints({
+          ...state.pagination,
+          pageSize,
+        }),
       }));
       forceUpdate();
-
-      const params = apiRef.current.getState<GridPaginationState>(
-        'pagination',
-      ) as GridPageChangeParams;
-      apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, {
-        // TODO remove params
-        ...params,
-        pageSize,
-      });
     },
-    [apiRef, setGridState, forceUpdate, logger, options.page],
+    [setGridState, forceUpdate, logger],
   );
 
-  useGridApiOptionHandler(apiRef, GRID_PAGE_CHANGE, options.onPageChange);
-  useGridApiOptionHandler(apiRef, GRID_PAGE_SIZE_CHANGE, options.onPageSizeChange);
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'page',
+      propModel: props.page,
+      propOnChange: props.onPageChange,
+      stateSelector: (state) => state.pagination.page,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.page, props.onPageChange]);
+
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'pageSize',
+      propModel: props.pageSize,
+      propOnChange: props.onPageSizeChange,
+      stateSelector: (state) => state.pagination.pageSize,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.pageSize, props.onPageSizeChange]);
+
+  React.useEffect(() => {
+    if (options.page === undefined) {
+      return;
+    }
+
+    setGridState((oldState) => ({
+      ...oldState,
+      pagination: {
+        ...oldState.pagination,
+        page: options.page!,
+      },
+    }));
+  }, [setGridState, options.page]);
```

Comment:
you can use the props directly

Reply:
That's a point I have some trouble understanding.
In the whole codebase, when should we use the props and when should we use the options ?

Reply:
we refactored hooks to take props as 2nd arguments so now you could use the props directly. Props are also now merged with the default props. So it's better to use props directly as it avoids a cycle to update options

Reply:
OK, I'm changing it for all the `options` in `useGridPagination` then

Reply:
https://app.circleci.com/pipelines/github/mui-org/material-ui-x/5958/workflows/d0ac1224-6900-4924-8cc6-be7dcaa96629/jobs/31174

This test fails if I use the props.
Before, the `useEfffect` in `useGridPagination` used `options.page` and `useGridSelector(apiRef, visibleGridRowCountSelector)` and both were always synchronized.
On the 1st run I have the default options page (0 page) and the default visible rows state (0 row).
O, the 2nd run I have the user options page (1 page) and the real visible rows state (1 row).

But now I'm using directly the prop.
So on the 1st run, I have the user options page (1 page) but I don't have the real visible rows state. I would have it if I ran the selector in the useEffect. But since it's in the render, it's still the value computed before `useGridFilter` updated the state.

```
1st render: Object{render: 0, effect: 3}
2nd render (caused by the initial `setGridState`: Object{render: 3, effect: 3}
3rd render (caused by the `setProps({ page: 2 })`) : Object{render: 3, effect: 3}
```